src/pass.h

```diff
@@ -191,6 +191,26 @@ struct PassOptions {
   // to the outside may be inspected in detail, which prevents us from e.g.
   // changing a type that escapes (so we can't remove or refine fields on an
   // escaping struct type, for example).
+  //
+  // Note that the module can still have imports and exports - otherwise it
+  // could do nothing at all! - so the meaning of "closed world" is a little
+  // subtle here. We do still want to keep imports and exports unchanged, as
+  // they form a contract with the outside world. For example, if an import has
+  // two parameters, we can't remove one of them. A nuance regarding that is how
+  // type equality works between wasm modules using the isorecursive type
+  // system: not only do we need to not remove a parameter as just mentioned,
+  // but we also want to keep types of things on the boundary unchanged. For
+  // example, we should not change an exported function's signature, as the
+  // outside may need that type to properly call the export.
+  //   * For now we disallow nontrivial types on the boundary, that is, you
+  //     cannot use a custom struct type as a function parameter type. Such a
+  //     type would not be optimizable due to the constraint to not modify the
+  //     type of the import/export, but it is very simple to use a type in a
+  //     single import/export and then all of its subtypes become unoptimizable,
+  //     which can mean all the types in your module if it is a root type such
+  //     as $java.lang.Object. For that reason we error on using such types on
```

Comment:
```suggestion
  //     single import/export and then all of its supertypes become unoptimizable,
  //     which can mean all the types in your module if it is a root type such
  //     as $java.lang.Object. For that reason we error on using such types on
```
The $java.lang.Object example would actually be fine. It's a type very deep in the hierarchy that might be a problem, since none of its supertypes would be able to change.

Reply:
It's both, isn't it? Imagine that we see that field `$hash` is never used anywhere in all of `$java.lang.Object` and everything else. If `$java.lang.Object` cannot be modified then GTO can't remove it, since subtypes can't have fewer fields.

Reply:
Whether a private subtype of a public supertype can be optimized depends on the optimization. Refining a field or removing a field not present in the supertype would be fine, for example. But a private supertype of a public subtype can never be optimized in any way because that would change the identity of the public type.

Reply:
Right, fair point, the main issue is that way.